src/fs_watcher.rs

```diff
@@ -0,0 +1,124 @@
+use crate::Dst;
+use crate::DstSender;
+use crate::EndpointMeta;
+use crate::Endpoints;
+use crate::Error;
+use inotify::{Event, EventMask, Inotify, WatchMask};
+use inotify_sys as ffi;
+use serde_json;
+use serde_yaml;
+use std::ffi::OsString;
+use std::mem;
+use std::string::String;
+use std::{fs, path::PathBuf};
+use tokio::stream::StreamExt;
+
+const EVENT_BUF_SZ: usize =
+    mem::size_of::<ffi::inotify_event>() + (libc::FILENAME_MAX as usize) + 1;
+
+#[derive(Debug)]
+pub struct FsWatcher {
+    endpoints_dir: PathBuf,
+    dst_sender: DstSender,
+}
+
+#[derive(Debug)]
+enum FileType {
+    Yaml,
+    Json,
+}
+
+#[derive(Debug)]
+pub struct FsWatcherError {
+    reason: String,
+}
+
+macro_rules! fs_watcher_error {
+    ($reason:expr) => {{
+        return Err(Box::new(FsWatcherError { reason: $reason }));
+    }};
+}
+
+// === impl FsWatcherError ===
+
+impl std::fmt::Display for FsWatcherError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        std::fmt::Display::fmt(&self.reason, f)
+    }
+}
+
+impl std::error::Error for FsWatcherError {}
+
+impl FsWatcher {
+    pub fn new(endpoints_dir: PathBuf, dst_sender: DstSender) -> Self {
+        Self {
+            endpoints_dir,
+            dst_sender,
+        }
+    }
+
+    fn parse_dst(file_name: &str) -> Result<(Dst, FileType), Error> {
+        let mut parts = file_name.rsplitn(2, ".");
+        match (parts.next(), parts.next()) {
+            (Some(ext), Some(name)) => {
+                let dst: Dst = name.parse()?;
+                let ft = match ext {
+                    "yaml" => FileType::Yaml,
```

Comment:
i think that `.yml` is a common enough extension for YAML files that we should probably recognize it here as well:
```suggestion
                    "yaml" | "yml" => FileType::Yaml,
```